### PR TITLE
SINGA-510 Add Time Profiling for Single GPU

### PR DESCRIPTION
**Pull request by chrishkchris**
Time Profiling for single GPU:
verbosity level: 0 (default) -> no logging
verbosity level: 1 -> forward and backward propagation time 
verbosity level: 2 -> each operation time (OP_ID, operation name, time)
The time is displayed as the average per iteration.

```
root@831960763a12:~/dcsysh/singa/examples/cnn# python3 train.py cnn mnist --verbosity 1
Starting Epoch 0:
Training loss = 578.907959, training accuracy = 0.796141
Evaluation accuracy = 0.937400, Elapsed Time = 5.622292s
Starting Epoch 1:
Training loss = 232.124695, training accuracy = 0.922609
Evaluation accuracy = 0.962841, Elapsed Time = 5.603526s
Starting Epoch 2:
Training loss = 167.437912, training accuracy = 0.944220
Evaluation accuracy = 0.971855, Elapsed Time = 5.422868s
Starting Epoch 3:
Training loss = 138.634125, training accuracy = 0.953392
Evaluation accuracy = 0.966747, Elapsed Time = 5.760310s
Starting Epoch 4:
Training loss = 117.458504, training accuracy = 0.961096
Evaluation accuracy = 0.973057, Elapsed Time = 5.756212s
Starting Epoch 5:
Training loss = 104.992790, training accuracy = 0.965198
Evaluation accuracy = 0.979267, Elapsed Time = 5.637915s
Starting Epoch 6:
Training loss = 96.263885, training accuracy = 0.967249
Evaluation accuracy = 0.980369, Elapsed Time = 5.753364s
Starting Epoch 7:
Training loss = 89.073364, training accuracy = 0.970051
Evaluation accuracy = 0.975561, Elapsed Time = 5.477089s
Starting Epoch 8:
Training loss = 82.311523, training accuracy = 0.972385
Evaluation accuracy = 0.980369, Elapsed Time = 4.668906s
Starting Epoch 9:
Training loss = 78.408806, training accuracy = 0.974270
Evaluation accuracy = 0.979968, Elapsed Time = 3.887881s

Time Profiling:
Forward Propagation Time : 0.00129212 sec
Backward Propagation Time : 0.00195529 sec
root@831960763a12:~/dcsysh/singa/examples/cnn# python3 train.py cnn mnist --verbosity 2
Starting Epoch 0:
Training loss = 578.907959, training accuracy = 0.796141
Evaluation accuracy = 0.937400, Elapsed Time = 4.777571s
Starting Epoch 1:
Training loss = 232.124695, training accuracy = 0.922609
Evaluation accuracy = 0.962841, Elapsed Time = 5.224710s
Starting Epoch 2:
Training loss = 167.437912, training accuracy = 0.944220
Evaluation accuracy = 0.971855, Elapsed Time = 5.266937s
Starting Epoch 3:
Training loss = 138.634125, training accuracy = 0.953392
Evaluation accuracy = 0.966747, Elapsed Time = 5.233795s
Starting Epoch 4:
Training loss = 117.458504, training accuracy = 0.961096
Evaluation accuracy = 0.973057, Elapsed Time = 5.114792s
Starting Epoch 5:
Training loss = 104.992790, training accuracy = 0.965198
Evaluation accuracy = 0.979267, Elapsed Time = 5.324732s
Starting Epoch 6:
Training loss = 96.263885, training accuracy = 0.967249
Evaluation accuracy = 0.980369, Elapsed Time = 5.413473s
Starting Epoch 7:
Training loss = 89.073364, training accuracy = 0.970051
Evaluation accuracy = 0.975561, Elapsed Time = 5.202558s
Starting Epoch 8:
Training loss = 82.311523, training accuracy = 0.972385
Evaluation accuracy = 0.980369, Elapsed Time = 5.240168s
Starting Epoch 9:
Training loss = 78.408806, training accuracy = 0.974270
Evaluation accuracy = 0.979968, Elapsed Time = 5.435664s

Time Profiling:
OP_ID0. cudnnConvForward : 0.000436663 sec
OP_ID1. cudnnAddTensor : 1.84035e-05 sec
OP_ID2. ReLU : 1.65394e-05 sec
OP_ID3. GpuPoolingForward : 1.69464e-05 sec
OP_ID4. cudnnConvForward : 7.17674e-05 sec
OP_ID5. cudnnAddTensor : 1.29785e-05 sec
OP_ID6. ReLU : 1.36852e-05 sec
OP_ID7. GpuPoolingForward : 1.5691e-05 sec
OP_ID8. GEMM : 3.60804e-05 sec
OP_ID9. SetValue : 0.00043388 sec
OP_ID10. GEMM : 2.06141e-05 sec
OP_ID11. ReLU : 1.20454e-05 sec
OP_ID12. GEMM : 2.70016e-05 sec
OP_ID13. SetValue : 0.000404889 sec
OP_ID14. GEMM : 1.56241e-05 sec
OP_ID16. SoftMax : 1.84867e-05 sec
OP_ID17. ComputeCrossEntropy : 1.24012e-05 sec
OP_ID18. SetValue : 7.12093e-05 sec
OP_ID19. SumAll : 2.48265e-05 sec
OP_ID20. Div : 1.02195e-05 sec
OP_ID22. CopyDataToFrom : 1.58324e-05 sec
OP_ID23. SoftmaxCrossEntropyBackward : 1.12932e-05 sec
OP_ID24. Div : 8.58751e-06 sec
OP_ID25. SetValue : 4.41526e-05 sec
OP_ID26. GEMV : 1.13555e-05 sec
OP_ID27. Axpy : 9.7881e-06 sec
OP_ID28. EltwiseMult : 2.51771e-05 sec
OP_ID29. Axpy : 8.55843e-06 sec
OP_ID30. Axpy : 9.03056e-06 sec
OP_ID31. GEMM : 1.95876e-05 sec
OP_ID32. GEMM : 1.71203e-05 sec
OP_ID33. Axpy : 9.33386e-06 sec
OP_ID34. EltwiseMult : 1.80905e-05 sec
OP_ID35. Axpy : 9.85235e-06 sec
OP_ID36. Axpy : 1.01951e-05 sec
OP_ID37. ReLUBackward : 1.13822e-05 sec
OP_ID38. SetValue : 1.89965e-05 sec
OP_ID39. GEMV : 1.08617e-05 sec
OP_ID40. Axpy : 9.05562e-06 sec
OP_ID41. EltwiseMult : 1.70098e-05 sec
OP_ID42. Axpy : 9.63604e-06 sec
OP_ID43. Axpy : 1.01054e-05 sec
OP_ID44. GEMM : 2.5537e-05 sec
OP_ID45. GEMM : 1.83094e-05 sec
OP_ID46. Axpy : 1.58145e-05 sec
OP_ID47. EltwiseMult : 2.04723e-05 sec
OP_ID48. Axpy : 1.6055e-05 sec
OP_ID49. Axpy : 2.054e-05 sec
OP_ID50. GpuPoolingBackward : 2.56164e-05 sec
OP_ID51. ReLUBackward : 1.4184e-05 sec
OP_ID52. cudnnConvolutionBackwardData : 7.3398e-05 sec
OP_ID53. cudnnConvolutionBackwardFilter : 7.19914e-05 sec
OP_ID54. cudnnConvolutionBackwardBias : 3.50412e-05 sec
OP_ID55. Axpy : 1.55906e-05 sec
OP_ID56. EltwiseMult : 1.38977e-05 sec
OP_ID57. Axpy : 1.45207e-05 sec
OP_ID58. Axpy : 6.90636e-05 sec
OP_ID59. Axpy : 1.37194e-05 sec
OP_ID60. EltwiseMult : 1.22707e-05 sec
OP_ID61. Axpy : 1.30436e-05 sec
OP_ID62. Axpy : 6.80907e-05 sec
OP_ID63. GpuPoolingBackward : 3.11083e-05 sec
OP_ID64. ReLUBackward : 3.09392e-05 sec
OP_ID65. cudnnConvolutionBackwardData : 6.54513e-05 sec
OP_ID66. cudnnConvolutionBackwardFilter : 9.62518e-05 sec
OP_ID67. cudnnConvolutionBackwardBias : 0.000128262 sec
OP_ID68. Axpy : 8.82783e-05 sec
OP_ID69. EltwiseMult : 1.14388e-05 sec
OP_ID70. Axpy : 4.30799e-05 sec
OP_ID71. Axpy : 1.65033e-05 sec
OP_ID72. Axpy : 7.64395e-05 sec
OP_ID73. EltwiseMult : 9.64545e-06 sec
OP_ID74. Axpy : 3.94519e-05 sec
OP_ID75. Axpy : 1.51809e-05 sec
```

### CODE REVIEW

**Comment by chrishkchris** [comment]
Time Profiling methods in NVIDIA documentation:
1. Chrono: https://docs.nvidia.com/deeplearning/sdk/tensorrt-best-practices/index.html#cpu-timing
2. CudaEvent: https://docs.nvidia.com/deeplearning/sdk/tensorrt-best-practices/index.html#cuda-events
I have implemented both.

File: include/singa/core/device.h [hunk 1]
```diff
@@ -107,12 +109,20 @@ class Device {
   Context* context(int k) { return &ctx_; }
 
   bool graph_enabled() const { return graph_enabled_; }
+  float verbosity() const { return verbosity_; }
```
Comment (nudles) [inline include/singa/core/device.h:112]: pls add some comments on the verbosity, e.g., what does 0, 1 and 2 mean.

**Comment by nudles** [comment]
@XJDKC can you help review this PR.

File: include/singa/core/device.h (hunk 1 above)
Comment (chrishkchris) [reply include/singa/core/device.h:112]: yes, I will add

**Comment by XJDKC** [comment]
> @XJDKC can you help review this PR.

Sure.

File: include/singa/core/device.h (hunk 1 above)
Comment (chrishkchris) [reply include/singa/core/device.h:117]: added, thanks

File: include/singa/core/device.h [hunk 2]
```diff
@@ -134,6 +150,8 @@ class Device {
   int num_executors_ = 0;
   unsigned seed_ = 0;
   bool graph_enabled_ = false;
+  string profiling_mode_ = "CHRONO";
+  int verbosity_ = 0;
```
Comment (XJDKC) [inline include/singa/core/device.h:154]: How about using two enum classes to distinguish the different profiling mode and the verbosity level?

File: src/core/scheduler/scheduler.cc [hunk 3]
```diff
@@ -228,6 +229,60 @@ void Graph::Debug() {
...
+    bool forward = true;
+    float forward_time = 0;
+    float backward_time = 0;
+    float time_elapsed;
+
+    for (size_t i = 0; i < nodes_.size(); ++i)
+      if (nodes_[i]->time_elapsed() > 0) {
+        if (forward == true)
+          // check the op of cross entropy backward, after that are backward ops
+          if (nodes_[i]->op_name().find("Backward") != std::string::npos)
```
Comment (XJDKC) [inline src/core/scheduler/scheduler.cc:246]: Do all backward propagation meet this condition?

File: src/model/layer/cudnn_activation.cc [hunk 4]
```diff
@@ -81,7 +81,7 @@ const Tensor CudnnActivation::Forward(int flag, const Tensor& input) {
     CUDNN_CHECK(cudnnActivationForward(
         ctx->cudnn_handle, this->acti_desc_, &alpha, this->desc_,
         inblock->data(), &beta, this->desc_, outblock->mutable_data()));
-  }, {input.block()}, {output.block()});
+  }, {input.block()}, {output.block()}, "cudnnActivationForward");
```
Comment (XJDKC) [inline src/model/layer/cudnn_activation.cc:84]: I haven't modified operations in files under the layer dir before, so they may be not well buffered.

File: include/singa/core/scheduler.h [hunk 5]
```diff
@@ -68,6 +76,11 @@ class Node {
   OpFunc op_;
   EdgeVec in_edges_;
   EdgeVec out_edges_;
+
+  string op_name_;
+  float time_elapsed_ = 0;
+  int iteration_ = 0;
```
Comment (XJDKC) [inline include/singa/core/scheduler.h:82]: I think we can move this variable to Graph. Although some ops might be executed more times than others when using old APIs.

File: src/core/scheduler/scheduler.cc (hunk 3 above)
Comment (chrishkchris) [reply src/core/scheduler/scheduler.cc:248]: I have added the note, it is more accurate when either "SoftmaxCrossEntropy" or "Softmax" is used, because both backward operation has "Backward" in the name as the dividing line between the forward and backward.

File: src/model/layer/cudnn_activation.cc (hunk 4 above)
Comment (chrishkchris) [reply src/model/layer/cudnn_activation.cc:84]: In my opinion, the folder src/model/layer is depecracted because it is not using the autograd API, so it has minimal value to be updated at this stage.

File: include/singa/core/scheduler.h (hunk 5 above)
Comment (chrishkchris) [reply include/singa/core/scheduler.h]: moved the variable to the graph, provide a step() function in graph to increment step counter

**Comment by XJDKC** [comment]
Ready for merge!

File: include/singa/core/device.h [hunk 6]
```diff
@@ -51,6 +51,17 @@ using std::vector;
 
 namespace singa {
 
+/// Method used by time profiling
+/// ProfilingMode == 0 (default) -> Chrono
+/// ProfilingMode == 1 -> Event (e.g. cudaEvent)
+enum ProfilingMode { useChrono, useEvent };
```
Comment (nudles) [inline include/singa/core/device.h:57]: I think chorno is only applicable for measuring cpu time and cudaevent is for measuring gpu time.
e.g., we cannot use cudaEvent if we are running on cpu.

File: include/singa/core/device.h [hunk 7]
```diff
@@ -51,6 +51,17 @@ using std::vector;
...
+
+/// Verbosity of the time profiling function:
+/// Verbosity == 0 (default) -> no logging
+/// Verbosity == 1 -> display forward and backward propagation time
+/// Verbosity == 2 -> display each operation time (OP_ID, op name, time)
+enum Verbosity { disableLog, forwardBackwardTime, eachOperation };
```
Comment (nudles) [inline include/singa/core/device.h:63]: I think verbosity is typically set as a number...
We can refer to other project's code on how to set verbosity, giving it a name or just using int.
logging has INFO, WARN, ERROR, which are self-explainable.

Comment (chrishkchris) [reply include/singa/core/device.h:63]: https://www.tensorflow.org/api_docs/python/tf/autograph/set_verbosity
Tensorflow is using number, so I will use number for verbosity

File: include/singa/core/device.h (hunk 6 above)
Comment (chrishkchris) [reply include/singa/core/device.h:57]: yes, cpu can use chrono only, gpu can use either 1. chrono or 2. cudaevent

Time Profiling methods in NVIDIA documentation:
1. Chrono: https://docs.nvidia.com/deeplearning/sdk/tensorrt-best-practices/index.html#cpu-timing
2. CudaEvent: https://docs.nvidia.com/deeplearning/sdk/tensorrt-best-practices/index.html#cuda-events
I have implemented both for GPU

Comment (nudles) [reply include/singa/core/device.h:57]: I think nvidia suggests to use chrono for cpu and use cudaevent for gpu operations because chorno may not measure the time correctly due to asynchronous execution; 
then we do not need users to select the mode.

Comment (chrishkchris) [reply include/singa/core/device.h:57]: ok, I delete the switch between the modes

Comment (chrishkchris) [reply include/singa/core/device.h:57]: I will modify the code to remove the switch

**Comment by chrishkchris** [comment]
I have addressed the review:
1. Use int for Verobsity
2. remove profiling mode (cpu use chrono, gpu use cudaevent)
It is ready for merge

File: include/singa/core/device.h (hunk 7 above)
Comment (chrishkchris) [reply include/singa/core/device.h]: I have modified the code: used int for Verbosity

File: include/singa/core/device.h (hunk 6 above)
Comment (chrishkchris) [reply include/singa/core/device.h]: I have removed the profiling mode, now cpu use chrono, gpu use cudaevent
